pkg/controller/cassandrarestore/cassandrarestore_controller.go

```diff
@@ -0,0 +1,396 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	AnnotationLastApplied string = "cassandrarestores.db.orange.com/last-applied-configuration"
+)
+
+var log = logf.Log.WithName("controller_cassandracluster")
+
+//var restoreFinalizer =  "finalizer.cassandrarestores.db.orange.com"
+
+// Add creates a new CassandraRestore Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileCassandraRestore{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	// Create a new controller
+	c, err := controller.New("cassandrarestore-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource CassandraCluster
+	err = c.Watch(&source.Kind{Type: &api.CassandraRestore{}}, &handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				object, err := meta.Accessor(e.Object)
+				if err != nil {
+					return false
+				}
+				if restore, ok := object.(*api.CassandraRestore); ok {
+					reqLogger := log.WithValues("Restore.Namespace", restore.Namespace, "Restore.Name", restore.Name)
+					cond := api.GetRestoreCondition(&restore.Status, api.RestoreRequired)
+					if cond != nil {
+						reqLogger.Info("Restore is already scheduled on Cluster member %s", restore.Spec.CoordinatorMember)
+						return false
+					}
+
+					return true
+				}
+				return false
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				object, err := meta.Accessor(e.ObjectNew)
+				if err != nil {
+					return false
+				}
+				if obj, ok := object.(*api.CassandraRestore); ok {
+					reqLogger := log.WithValues("Restore.Namespace", obj.Namespace, "Restore.Name", obj.Name)
+					//old := e.ObjectOld.(*api.CassandraRestore)
+					new := e.ObjectNew.(*api.CassandraRestore)
+					if new.Status.Condition == nil {
+						return false
+					}
+					if new.Status.Condition.Type.IsCompleted() {
+						reqLogger.Info("Restore is completed, skipping.")
+						return false
+					}
+
+					if new.Status.Condition.Type.IsInError() {
+						reqLogger.Info("Restore is in error state, skipping.")
+						return false
+					}
+
+					return true
```

Comment:
When is that condition met ? I'm not sure to understand

Reply:
When the restore is neither completed nor in error